aspnetcore/data/ef-rp/complex-data-model.md

```diff
@@ -0,0 +1,643 @@
+
+## Update the DB context
+
+Add the following highlighted code to *Data/SchoolContext.cs*:
+
+[!code-csharp[Main](intro/samples/cu/Data/SchoolContext.cs?name=snippet_BeforeInheritance&highlight=15-18,25-31)]
+
+The preceding code adds the new entities and configures the `CourseAssignment` entity's composite PK.
+
+## Fluent API alternative to attributes
+
+The `OnModelCreating` method in the preceding code uses the *fluent API* to configure EF behavior. The API is called "fluent" because it's often used by stringing a series of method calls together into a single statement. The [following code](https://docs.microsoft.com/ef/core/modeling/#methods-of-configuration) is an example of the fluent API:
+
+```csharp
+protected override void OnModelCreating(ModelBuilder modelBuilder)
+{
+    modelBuilder.Entity<Blog>()
+        .Property(b => b.Url)
+        .IsRequired();
+}
+```
+
+In this tutorial, the fluent API is used only for DB mapping that can't be done with attributes. However, the fluent API can specify most of the formatting, validation, and mapping rules that can be done with attributes.
+
+Some attributes such as `MinimumLength` can't be applied with the fluent API. `MinimumLength` doesn't change the schema, it only applies a minimum length validation rule.
+
+Some developers prefer to use the fluent API exclusively so that they can keep their entity classes "clean." Attributes and the fluent API can be mixed. There are some configurations that can only be done with the fluent API (specifying a composite PK). There are some configurations that can only be done with attributes (`MinimumLength`). The recommended practice for using fluent API or attributes:
+
+* Choose one of these two approaches.
+* Use the chosen approach consistently as much as possible.
+
+For more information about attributes vs. fluent API, see [Methods of configuration](https://docs.microsoft.com/ef/core/modeling/#methods-of-configuration).
+
+## Entity Diagram Showing Relationships
+
+The following illustration shows the diagram that EF Power Tools create for the completed School model.
+
+![Entity diagram](complex-data-model/_static/diagram.png)
+
+The preceding diagram shows:
+
+* Several one-to-many relationship lines (1 to \*).
+* The one-to-zero-or-one relationship line (1 to 0..1) between the `Instructor` and `OfficeAssignment` entities.
+* The zero-or-one-to-many relationship line (0..1 to *) between the `Instructor` and `Department` entities.
+
+## Seed the DB with Test Data
+
+Update the code in *Data/DbInitializer.cs*:
+
+[!code-csharp[Main](intro/samples/cu/Data/DbInitializer.cs?name=snippet_Final)]
+
+The preceding code provides seed data for the new entities. Most of this code creates new entity objects and loads sample data. The sample data is used for testing. The preceding code creates the following many-to-many relationships:
```

Comment:
For 2.1 this should be changed to use the seeding mechanism built into EF and migrations.